 ╔══<i><b>Alai-DeepLearning</b></i>════════════════════════╗
###  &nbsp;&nbsp; **✎&nbsp;&nbsp;Week 2. Numpy Basis**
# Section 2. Numpy의 구조와 생성

### _Objective_
1. numpy array의 속성인 shape, dtype에 대해 배웁니다. <br>
2. numpy array를 만드는 다양한 방식에 대해 배웁니다. <br>

╚═════════════════════════════════════╝

In [0]:
import numpy as np

# \[ 1. Numpy Array의 구조 \]

----

----

> *Numpy Array는 Value(element), Shape, Datatype으로 구성되어있습니다.*<br>
> *Numpy Array의 중요한 특징인 shape과 datatype에 대해 배워봅시다.*

<br>
## 1. Shape

---

+ Numpy에서의 Shape은 차원 array의 모양, 즉 각 차원별 요소의 수를 의미합니다.<br>
+ 아래 예제를 통해 "차원"과 "shape"의 개념을 익혀보겠습니다.


### (1) 1차원 행렬 생성


1차원 행렬은 주로 Vector(벡터)라 불립니다.
<br>

In [0]:
A = np.array([1,2,3,4,5])
A

array([1, 2, 3, 4, 5])

In [0]:
A.ndim # 차원 수 

1

In [0]:
A.shape # array의 형태

(5,)

In [0]:
A.size # array 내부의 원소 수  == 5

5

### (2) 2차원 행렬 생성

2차원 행렬은 수학에서 주로 Matrix(행렬)이라 불립니다.
<br>

In [0]:
A = np.array([
    [1,2,3,4,5],
    [2,3,4,5,6],
    [3,4,5,6,7]
],dtype=np.int64)
A

array([[1, 2, 3, 4, 5],
       [2, 3, 4, 5, 6],
       [3, 4, 5, 6, 7]])

In [0]:
A.ndim # 차원 수 

2

In [0]:
A.shape # 행이 3개, 열이 5개

(3, 5)

In [0]:
A.size # 3(행의 수) * 5(열의 수) == 15

15

### ⚠︎ 주의 : Python의 `list`와 다르게, Numpy의 array는 행, 열 별로 크기가 같아야 합니다.
╭──────────────────────────────────────────────────────────╮

> `list`는 아래와 같이 가변적인 길이를 담을 수 있습니다.

In [0]:
# list는 아래와 같이 담을 수 있지만,
A_ = [
    [1,2,3,4,5], # 크기가 5
    [2,3,4,5], # 크기가 4
    [3,4,5,6,7] # 크기가 5
]
A_

[[1, 2, 3, 4, 5], [2, 3, 4, 5], [3, 4, 5, 6, 7]]

> `numpy.array`는 아래와 같이 각 행, 열별로 크기가 같지 않으면, 넣을 수 없습니다.

In [0]:
# array는  담을 수 없습니다.
A = np.array([
    [1,2,3,4,5], # 크기가 5
    [2,3,4,5], # 크기가 4
    [3,4,5,6,7] # 크기가 5
],dtype=np.int64)
A

ValueError: setting an array element with a sequence.

> `np.array`는 수학에서의 행렬과 같이, 행 별, 열 별로 같은 갯수만큼의 원소가 있어야 합니다.

###  
╰──────────────────────────────────────────────────────────╯

### (3) 3차원 행렬 생성

3차원이상의 행렬은 수학에서 주로 Multidimensional Matrix라 불립니다.<br>
이렇게 크기가 N차원인 모든 행렬들을 통칭해 Tensor라 부르고, <br>
이후 배우는 `tensorflow` 라이브러리는 이러한 `tensor`을 연산하기 위해 고안된 라이브러리입니다.<br>

In [0]:
A = np.array([
    [
        [1,2,3,4,5],
        [2,3,4,5,6],
        [3,4,5,6,7]
    ],
    [
        [1,2,3,4,5],
        [2,3,4,5,6],
        [3,4,5,6,7]
    ]
])
A

array([[[1, 2, 3, 4, 5],
        [2, 3, 4, 5, 6],
        [3, 4, 5, 6, 7]],

       [[1, 2, 3, 4, 5],
        [2, 3, 4, 5, 6],
        [3, 4, 5, 6, 7]]])

In [0]:
A.ndim # 차원 수 

3

In [0]:
A.shape # 0번째 차원이 2, 1번째 차원이 3, 2번째 차원이 5

(2, 3, 5)

In [0]:
A.size  # 2(0번째 차원) x 3(1번째 차원) x 5(2번째 차원)

30

### (4) shape 변경하기, `reshape`

원소의 수가 같은 경우, `numpy array`의 형태를 바꿀 수 있습니다.
<br>

In [0]:
A = np.array([
    [
        [1,2,3,4,5],
        [2,3,4,5,6],
        [3,4,5,6,7]
    ],
    [
        [1,2,3,4,5],
        [2,3,4,5,6],
        [3,4,5,6,7]
    ]
])
A

array([[[1, 2, 3, 4, 5],
        [2, 3, 4, 5, 6],
        [3, 4, 5, 6, 7]],

       [[1, 2, 3, 4, 5],
        [2, 3, 4, 5, 6],
        [3, 4, 5, 6, 7]]])

In [0]:
A.reshape((3,2,5)) 

array([[[1, 2, 3, 4, 5],
        [2, 3, 4, 5, 6]],

       [[3, 4, 5, 6, 7],
        [1, 2, 3, 4, 5]],

       [[2, 3, 4, 5, 6],
        [3, 4, 5, 6, 7]]])

`reshape`을 통해, 행렬의 형태를 바꿀 수 있습니다.<br>
이러한 행렬의 shape 변환은 아래의 예제에서 유용하게 사용할 수 있습니다.
<br>

#### 예제 )  제품별 월 판매량 정보

┌─<i> ✍︎ Example  </i>────────────────────────────┐ <br>

&nbsp;&nbsp;&nbsp;&nbsp;제품별 **월 판매량** 정보데이터에서 **분기별 판매량** 정보데이터로 변환하기<br>

└───────────────────────────────────┘<br>

|월별 | 1 | 2 | 3 | 4 | 5 |6|7|8|9|10|11|12|
|---|----|---|---|---|---|---|---|---|---|---|---|
|A모델|30|33|35|31|36|32|37|39|42|37|38|39|
|B모델|20|22|25|21|23|29|32|30|35|31|36|38|
|C모델|10|12|11|13|10|16|12|11|15|12|13|12|

> 위의 데이터 셋이 있을 때, **분기별** 판매량을 구하고 싶다면, 어떻게 해야할까?

In [0]:
sales = np.array([
    [30,33,35,31,36,32,37,39,42,37,38,39],
    [20,22,25,21,23,29,32,30,35,31,36,38],
    [10,12,11,13,10,16,12,11,15,12,13,12]
])
sales.shape # shape = (모델 별(a,b,c), 월 별 (1~12월))

(3, 12)

|분기 | 1Q(1~3) | 2Q(4~6) | 3Q(7~9) | 4Q(10~12) |
|---|----|---|---|---|
|A모델|[30,33,35]|[31,36,32]|[37,39,42]|[37,38,39]|
|B모델|[20,22,25]|[21,23,29]|[32,30,35]|[31,36,38]|
|C모델|[10,12,11]|[13,10,16]|[12,11,15]|[12,13,12]|

위와 같이 shape을 바꾼 후, 각 분기별로 합을 구하는 방식으로 진행하면 간편합니다.
<br>

In [0]:
sales = sales.reshape(3,4,3) # (모델 별, 분기 별, 분기 내 월 별) == (3, 4, 3)
sales

array([[[30, 33, 35],
        [31, 36, 32],
        [37, 39, 42],
        [37, 38, 39]],

       [[20, 22, 25],
        [21, 23, 29],
        [32, 30, 35],
        [31, 36, 38]],

       [[10, 12, 11],
        [13, 10, 16],
        [12, 11, 15],
        [12, 13, 12]]])

In [0]:
sales.sum(axis=2)

array([[ 98,  99, 118, 114],
       [ 67,  73,  97, 105],
       [ 33,  39,  38,  37]])

#### 결과) 분기별 판매량 정보

|분기 | 1Q | 2Q | 3Q | 4Q |
|---|----|---|---|---|
|A모델|98|99|118|114|
|B모델|67|73|97|105|
|C모델|33|39|38|37|

> * reshape을 통해, 간단히 분기별 판매량을 구할 수 있습니다. 

<br>

## 2. DataType

<hr>

+ Python에서는 Datatype을 데이터에 따라 자동으로 설정됩니다.
+ 하지만 Numpy의 Numpy Array에서는 Datatype이 모두 동일해야 하기 때문에, Datatype을 확인하거나 Datatype을 변환해야 할때도 있습니다.

### (1) `list`와 `np.array`의 차이, datatype

`np.array`는 `list`와 다르게 행렬 내 모든 타입이 같은 타입이어야만 합니다.

In [0]:
# List
A_ = [1,1.2,False]
A_

[1, 1.2, False]

List는 원소 별로 Type이 달라도 됩니다.

In [0]:
# numpy array
A = np.array([1,1.2,False])
A

array([1. , 1.2, 0. ])

`1`과 `True`가 각각 `1.`, `0.`으로 바뀝니다. **이것을 형변환(type-casting)이라고 합니다.**

### (2) `datatype` 확인하기

#### &nbsp;&nbsp;&nbsp;&nbsp; ⓵ boolean array

In [0]:
A = np.array([True,False])
A

array([ True, False])

In [0]:
A.dtype

dtype('bool')

> * `dtype`을 통해, 현재 `np.array`의 데이터 타입이 어떤 형태인지 알 수 있습니다.

#### &nbsp;&nbsp;&nbsp;&nbsp; ⓶ int array

In [0]:
A = np.array([1,2,3])
A

array([1, 2, 3])

In [0]:
A.dtype

dtype('int64')

#### &nbsp;&nbsp;&nbsp;&nbsp; ⓷ float array

In [0]:
A = np.array([1.0,2.0,3.0])
A

array([1., 2., 3.])

In [0]:
A.dtype

dtype('float64')

#### ※ numpy의 다양한 데이터 타입들


|Data Type | Description |
|----|----|
|bool_ | Boolean (True or False) stored as a byte|
|int_ | Default integer type (same as C long; normally either int64 or int32)|
|intc | Identical to C int (normally int32 or int64)|
|intp | Integer used for indexing (same as C ssize_t; normally either int32 or int64)|
|int8 | Byte (-128 to 127)|
|int16 | Integer (-32768 to 32767)|
|int32 | Integer (-2147483648 to 2147483647)|
|int64 | Integer (-9223372036854775808 to 9223372036854775807)|
|uint8 | Unsigned integer (0 to 255)|
|uint16 | Unsigned integer (0 to 65535)|
|uint32 | Unsigned integer (0 to 4294967295)|
|uint64 | Unsigned integer (0 to 18446744073709551615)|
|float_ | Shorthand for float64.|
|float16 | Half precision float: sign bit, 5 bits exponent, 10 bits mantissa|
|float32 | Single precision float: sign bit, 8 bits exponent, 23 bits mantissa|
|float64 | Double precision float: sign bit, 11 bits exponent, 52 bits mantissa|
|complex_ | Shorthand for complex128.|
|complex64 | Complex number, represented by two 32-bit floats (real and imaginary components)|
|complex128 | Complex number, represented by two 64-bit floats (real and imaginary components)|

### (3) `astype`으로 형 변환하기

In [0]:
A = np.array([0,1,2,3,4,5])
A

array([0, 1, 2, 3, 4, 5])

`astype`으로, 행렬의 datatype을 쉽게 바꿀 수 있습니다.

In [0]:
A.astype(np.float) # int -> float으로 바꾸기

array([0., 1., 2., 3., 4., 5.])

In [0]:
A.astype(np.bool) # int -> bool으로 바꾸기

array([False,  True,  True,  True,  True,  True])

<br><br>

# \[ 2. Numpy Array을 생성하기 \]

----

---- 
> *Numpy Array을 만드는 다양한 방식을 배워봅시다.*

<br>
## 1. 다른 자료로부터 Numpy Array 생성하기

---

* 우리는 이미 생성된 List나 Tuple 또는 또다른 Array를 활용하여 새로운 Array를 생성할 수 있습니다.<br>
* List, Tuple로부터 Numpy Array를 생성하는 `np.array` 와 다른 Numpy Array로부터 Numpy Array를 생성하는 `np.copy` 를 활용해봅시다.


### (1) 리스트를 Array로, `np.array()`

가장 기본적인 array를 만드는 방식으로, 주로 리스트나 tuple을 변환하여 array로 만들어 줍니다.

In [0]:
A = np.array([1,2,3,4,5])
A

array([1, 2, 3, 4, 5])

### (2) Array을 복사하기, `np.copy()`

똑같은 값을 가진 Array를 만들고자 한다면, `np.copy()`를 이용해야 합니다.

In [0]:
A = np.array([1,2,3,4,5])
A

array([1, 2, 3, 4, 5])

In [0]:
B = A # 이렇게 할 경우, A를 복사하는 것이 아닌, 똑같은 A의 값을 가르키고 있는 상태가 됩니다.

In [0]:
B[0] = 0 # B의 첫번째 원소를 바꾸었을 때, A도 바뀝니다.
B

array([0, 2, 3, 4, 5])

In [0]:
A # B와 같음

array([0, 2, 3, 4, 5])

<br>
위와 같은 것을 방지하기 위해서는, `copy`라는 메소드를 이용해야 합니다.
<br>

In [0]:
A = np.array([1,2,3,4,5])
A

array([1, 2, 3, 4, 5])

In [0]:
B = A.copy() # 이렇게 할 경우, A를 복사하여 저장합니다.

In [0]:
B[0] = 0 # B의 첫번째 원소를 바꾸었을 때, A는 바뀌지 않습니다.
B

array([0, 2, 3, 4, 5])

In [0]:
A # B와 다름

array([1, 2, 3, 4, 5])

<br>

## 2. 구성을 설정하여 Numpy Array 생성하기

<hr>

우리는 구성(Value, Shape, Datatype)을 설정하여 Numpy Array를 직접 생성할수도 있습니다.<br>
+ 모든 요소(Elements)가 동일한 값(Value)를 가지는 Numpy Array를 생성할 수도 있으며,<br>
+ Shape를 지정하거나 다른 Numpy Array와 같은 Shape를 가지는 Numpy Array를 생성할 수도 있고,<br>
+ 요소들의 값들을 등간격으로 설정하여 Numpy Array를 생성할 수도 있습니다.


### (1) 값이 일정한 Array&nbsp;,&nbsp; `np.zeros()`&nbsp;,&nbsp; `np.ones()`&nbsp;,&nbsp; `np.full()`

우리는 Array를 지정할 때 크게 `shape`, `value`, `dtype`을 지정해야 합니다.

In [0]:
A = np.zeros(shape=(3,2), dtype=np.float64)
A

array([[0., 0.],
       [0., 0.],
       [0., 0.]])

In [0]:
A = np.ones(shape=(3,2), dtype=np.float64)
A

array([[1., 1.],
       [1., 1.],
       [1., 1.]])

In [0]:
A = np.full(shape=(3,2), fill_value=3., dtype=np.float64)
A

array([[3., 3.],
       [3., 3.],
       [3., 3.]])

### (2) 똑같은 Shape의 Array, &nbsp; `np.zeros_like()`&nbsp;, &nbsp;`np.ones_like()`&nbsp;, &nbsp;`np.full_like`&nbsp;
다른 Numpy Array의 Shape와 동일한 Shape를 갖는 Numpy Array를 생성할 수 있습니다.

In [0]:
# Shape를 본 딸 Numpy Array : A 생성
A = np.array([1,2,3,4,5])
A

array([1, 2, 3, 4, 5])

In [0]:
np.zeros_like(A, dtype=np.int) # 1. A와 같은 Shape이면서 모든 요소의 값이 0 인 Numpy Array 생성

array([0, 0, 0, 0, 0])

In [0]:
np.ones_like(A, dtype=np.int) # 2. A와 같은 Shape이면서 모든 요소의 값이 1 인 Numpy Array 생성 

array([1, 1, 1, 1, 1])

In [0]:
np.full_like(A, fill_value=13, dtype=np.int) # 3. A와 같은 Shape이면서 모든 요소의 값이 13 인 Numpy Array 생성

array([13, 13, 13, 13, 13])

### (3) 같은 간격의 Array, &nbsp; `np.linspace()`&nbsp; , &nbsp; `np.arange()`

[1,2,3,4,5,6,7,8,9,10,...] 이런 식으로 등간격으로 배치된 Array를 만들고 싶을 때, 우리는 간단히 위의 메소드를 통해 만들 수 있습니다.

In [0]:
np.linspace(start=1,stop=10, num=19) # 1부터 10까지 19개의 요소를 등간격으로 쪼개서 생성

array([ 1. ,  1.5,  2. ,  2.5,  3. ,  3.5,  4. ,  4.5,  5. ,  5.5,  6. ,
        6.5,  7. ,  7.5,  8. ,  8.5,  9. ,  9.5, 10. ])

In [0]:
np.arange(1,10,0.5) # 1부터 10까지 0.5 step(간격)으로 더하여 요소를 생성

array([1. , 1.5, 2. , 2.5, 3. , 3.5, 4. , 4.5, 5. , 5.5, 6. , 6.5, 7. ,
       7.5, 8. , 8.5, 9. , 9.5])

* `np.linspace`는 **<몇 개의 숫자로 등간격으로 나눌까?>를 기준**으로 array를 만들고,
* `np.arange`는 **<간격을 얼마로 나눌까?>를 기준**으로 array를 만듭니다.**

#  

---

    Copyright(c) 2019 by Public AI. All rights reserved.
    Writen by PAI, SangJae Kang ( rocketgrowthsj@publicai.co.kr )  last updated on 2019/03/04


---